fe_contracts/defi/src/raw_call.fe

```diff
@@ -21,40 +21,42 @@ pub struct RawCallHelper {
 
     /// Call a function directly by its selector on the given address. 
     pub fn call<T: MemoryEncoding>(_ addr: address, _ selector: u32, data: T) -> RawCallResult{
+        let mut start: u256
         unsafe {
             let input_offset: u256 = RawCallHelper::avail()
             [selector]
             data.write_mem()
-
+            start = RawCallHelper::avail() + data.byte_count()
             let success: u256 = std::evm::call(
```

Comment:
Does `write_mem` update the free memory pointer? It doesn't appear so. 

Also, I tried something like to update the free memory pointer but it didn't seem to work:
```
std::evm::mstore(offset: 64, value: input_offset + data.byte_count())
start = RawCallHelper::avail()
```
It didn't appear to succeed but `start = RawCallHelper::avail() + data.byte_count()` does. Am I missing something?

Reply:
>Does `write_mem` update the free memory pointer? It doesn't appear so.

It does, but it doesn't do so explicitly. Please note that the whole thing is really just a shitty crux because I haven't found any cleaner way to handle externals for when Fe can not match the selector natively.

In current Fe, there's no way to express bytes without giving it an explicit size and hence the only way I've found to make a somewhat workable abstraction was to create the `MemoryEncoding` trait and leave it up to the implementer to write the data to memory. 

In this case, I just call `abi_encode`, knowing that this will write the data to memory so that we can pick it up from there.

``` 
    fn write_mem(self) {
        self.abi_encode()
    }
```
It's extremely hacky and fragile to rely on this. We can take a look at the underlying YUL if we run `$FE_PATH build --overwrite fe_contracts/defi/ --emit yul`.

We can see the underlying YUL code for this particular piece of code:

```
            function main$MemoryEncoding$ExactInputSingleParams$write_mem($self) {
                let $$tmp_1 := $$alloc(256)
                pop($$abi_encode_static_aggregate_type_20_to_memory($self, $$tmp_1))
                leave
            }
```

The `alloc(256)` does in fact update the free memory pointer.

Reply:
I think the issue may be related to scoping. The `start` declared in safe Fe isn't updated when it's assigned in `unsafe` Fe. 
```
function raw_call$RawCallHelper$call_ExactInputSingleParams($addr, $selector, $data) -> $$ret {
      let $$tmp_3 // let mut start: u256
      let $$tmp_4
      $$tmp_4 := and(raw_call$RawCallHelper$avail(), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
      let $$tmp_5 := $$alloc(4)
      $aggregate_init_34($$tmp_5, $selector)
      main$MemoryEncoding$ExactInputSingleParams$write_mem($data)
      let $$tmp_7 // start (unsafe scope)
      $$tmp_7 := and(raw_call$RawCallHelper$avail(), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
      let $$tmp_8
      $$tmp_8 := and(main$MemoryEncoding$ExactInputSingleParams$byte_count($data), 0xffff)
      let $$tmp_9
      $$tmp_9 := and(and($$tmp_8, 0xffff), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
      $$tmp_3 := and($$safe_add_unsigned($$tmp_7, $$tmp_9, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
      let $$tmp_10
      let $$tmp_11
      $$tmp_11 := and(evm$gas_remaining(), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
      let $$tmp_14
      $$tmp_14 := and(main$MemoryEncoding$ExactInputSingleParams$byte_count($data), 0xffff)
      let $$tmp_15
      $$tmp_15 := and(and($$tmp_14, 0xffff), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
      let $$tmp_16
      $$tmp_16 := and($$safe_add_unsigned(4, $$tmp_15, 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff)
      $$tmp_10 := and(evm$call($$tmp_11, $addr, 0, $$tmp_4, $$tmp_16, $$tmp_3, 32), 0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff) 
```
`evm$call($$tmp_11, $addr, 0, $$tmp_4, $$tmp_16, $$tmp_3, 32)` use `$$tmp_3` instead of `$$tmp_7`